neuralmonkey/encoders/recurrent.py

```diff
@@ -178,8 +167,41 @@ def rnn_input(self) -> tf.Tensor:
 
     @tensor
     def rnn(self) -> Tuple[tf.Tensor, tf.Tensor]:
-        return rnn_layer(self.rnn_input, self.input_sequence.lengths,
-                         self.rnn_spec, self.add_residual)
+        layer_input = self.rnn_input  # type: tf.Tensor
+        # pylint: disable=unsubscriptable-object
+        layer_final = self.rnn_input[:, -1]
+        # pylint: enable=unsubscriptable-object
+
+        for i, rnn_spec in enumerate(self.rnn_specs):
+            with tf.variable_scope("rnn_{}_{}".format(i, rnn_spec.direction),
+                                   reuse=tf.AUTO_REUSE):
+
+                if self.add_layer_norm:
+                    layer_input = layer_norm(layer_input)
+
+                layer_output, layer_final_output = rnn_layer(
+                    layer_input, self.input_sequence.lengths, rnn_spec)
+
+                layer_output = dropout(
+                    layer_output, self.dropout_keep_prob, self.train_mode)
+                layer_final_output = dropout(
+                    layer_final_output, self.dropout_keep_prob,
+                    self.train_mode)
+
+                in_dim = layer_input.get_shape()[-1]
+                out_dim = layer_output.get_shape()[-1]
+
+                if self.add_residual and in_dim == out_dim:
+                    layer_input += layer_output
+                    layer_final += layer_final_output
+                else:
+                    # pylint: disable=redefined-variable-type
+                    layer_input = layer_output
+                    layer_final = layer_final_output
+                    # pylint: enable=redefined-variable-type
+
+        assert layer_final is not None
+        return layer_input, layer_final
```

Comment:
Na vystup bych taky pridal layer_norm

Reply:
Nematus to dělá na vstupu (https://github.com/EdinburghNLP/nematus/blob/master/nematus/layers.py)
Tensor2tensor to taky dělá na vstupu
(https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py#L1618)

Dává smysl se nad normalizací zamýšlet jako na preprocessingu toho, co ta vrstva dělá, protože layer normalizaci aplikuju kvůli nějaký stabilitě toho, co chci teď počítat. Nad dropoutem můžeš dál uvažovat jako na postprocessingu, stejně jako nad residual connectionem.

Reply:
Transformer to dela sice na vstupu, ale na finalni vystup enkoderu se pote take aplikuje layer_norm, viz:
https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/transformer_layers.py#L223
a nebo nase implementace:
https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/encoders/transformer.py#L312

Kod Nematuse neznam, ale predpokladam, ze taky vystupni stavy enkoderu layer normalizuji, nez je nacpou do attentionu.

Navic, kdybychom, ze v tom chteli skutecne vrtat, v GNMT pouzivaj layer normalizovanou LSTM bunku, (https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LayerNormBasicLSTMCell), ktera na prvni pohled dela sofistikovanejsi layer_norm primo na jednotlivych gejtach a ne jen na vystupech vrstvy:
https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/contrib/rnn/python/ops/rnn_cell.py#L1445
(musel bych si to ale procist poradneji, abych to rekl s jistotou)

Reply:
gnmt má podle mě layer norm naimplementovaný stejně jak nematus. no, tohle je jedna z těch věcí, kde bude fungovat obojí a dokud se nezměří, co je lepší, tak se o tom dá polemizovat do aleluja..

Reply:
No docela by me zajimalo, jaka je logika za tim, normalizovat temp. states mezi kazdou vrstvou (a samozrejme i vstupni embedded tokeny), ale pak z nejakeho duvodu temp. states, co vylezou z enkoderu nenormalizovat.

Myslim, ze kdyby ses vic vrtal v nematovi, tak bys zjistil, ze ty vystupy enkoderu normalizuji, ale treba az kdyz nad nimi pocitaji attention (takze pouze v jine casti kodu). Uz jen to, ze ty stavy v attentionu vazene scitas motivuje k tomu, aby jejich hodnoty mely nejakou rozumnou stredni hodnotu a rozptyl.

Reply:
Moc se mi ty layer normy nechce předělávat, protože na tomhle kódu už měsíc jedou language modely na náš semisupervised translation a jakmile to změní, tak modely, co takhle vytvoříme už nikdy nepustíme na masteru. A ELMo, podle kterého jsem dělal ty LM, taky dělá layer norm na vstupu.

Reply:
A pred softmaxem layer_norm nedelas?

Reply:
Nedělám

Reply:
> Moc se mi ty layer normy nechce předělávat, protože na tomhle kódu už měsíc jedou language modely na náš semisupervised translation a jakmile to změní, tak modely, co takhle vytvoříme už nikdy nepustíme na masteru. A ELMo, podle kterého jsem dělal ty LM, taky dělá layer norm na vstupu.

Podarilo se ti nacist predtrenovany ELMo model v opici (tj., skutecne nam nechybi nektere parametry v nasi implementaci)? Stale mi prijde divne, ze ten layer_norm neaplikujes na vystup RNN pred FF na velikost slovniku (a naslednym softmaxem). Vzhledem k tomu, ze layer_norm by mel fungovat i s FF vrstvami (viz. transformer)

Tohle je IMO podporeno i v orinal paperu na Layer Normalisation:
"With fully connected layers, all the hidden units in a layer tend to make similar contributions to the final prediction and re-centering and rescaling the summed inputs to a layer works well."

Reply:
ELMo jsem ani neměl snahu načítat. Ta layer norm by tam potom měla bejt (a já jí tam nemám), ale to by si měl šéfovat už labeler, aby stavy, co bych potom použil jako reprezentaci jinde nebyly zatížený tím, že po sloupečkách sčítají do stejného čísla, ne?

Reply:
Osobne si myslim, ze layer_normalizace nema moc cemu uskodit. Spis se nabizi podobna otazka ala dropout, jestli delat norm. na vstup nebo na vystup.
Zkousel jsem si narychlo odvodit, co se stane, kdyz aplikujes layer_norm dvakrat po sobe a vychazi mi, ze by v tom nemel byt rozdil, pokud oba layer_normy budou mit stejny gain a bias (coz se snad backpropem nauci).

A kdyz uz se v tom tak vrtame:
Vubec nedelame layer_norm na hidden state transition, delas ho pouze na vstup. V tom layer_norm paperu normalizuji soucet linearnich projekci vstupu a hidden statu. Otazka ja, jak moc je tohle nezbytne.

Reply:
Ok, tak já udělám přepínač, jestli normalizovat poslední vrstvu, co bude faultně na true.

Reply:
Ok, to zni jako dobry kompromis :)